xwiki-platform-core/xwiki-platform-livedata/xwiki-platform-livedata-webjar/src/main/webjar/Logic.js

```diff
@@ -0,0 +1,119 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+
+define([], function () {
+
+    /**
+     * Map the element to its data object
+     * So that each instance of the livedata on the page handle there own data
+     */
+    var instancesMap = new WeakMap();
+
+
+
+    /**
+     * The init function of the logic script
+     * For each livedata element on the page, returns its corresponding data / API
+     * If the data does not exists yet, create it from the element
+     * @param {HTMLElement} element The HTML Element corresponding to the Livedata component
+     */
+    var init = function (element) {
+
+        if (!instancesMap.has(element)) {
+            // create a new logic object associated to the element
+            var logic = new Logic(element);
+            instancesMap.set(element, logic);
+
+            logic.loadLayout();
+        }
+
+        return instancesMap.get(element);
+    };
+
+
+    /**
+     * Class for a logic element
+     * Contains the Livedata data object and methods to mutate it
+     * Can be used in the layouts to display the data, and call its API
+     * @param {HTMLElement} element The HTML Element corresponding to the Livedata
+     */
+    var Logic = function (element) {
+        this.element = element;
+        this.data = JSON.parse(element.getAttribute("data-data") || "{}");
+        element.removeAttribute("data-data");
+        this.layouts = {};
+    };
+
+
+    /**
+     * Load a layout, or default layout if none specified
+     * @param {String} layoutName The name of the layout to load with requireJS
+     */
+    Logic.prototype.loadLayout = function (layoutName) {
+        var self = this;
+
+        layoutName = layoutName || this.data.layouts["default"];
```

Comment:
I don't think we need to have a default layout explicitly specified with a dedicated ``default`` property. I usually do this when reading some config:

    actualConfig = $.extend(true, {}, defaultConfig, configReadFromElement)

Having ``default`` properties for any configuration option is too much for me.

Reply:
No no, I didn't mean to create a default for any configuration option, the `layouts.default` was just in case there are several layouts available, and we want to specify which one should be displayed by default to the user.
If you don't want to have a specific property for this though, we can also say that the first layout in the array of available layouts should be used by default, but this might be less intuitive. Or a mix of both: if default layout is not specified, it takes the first available layout.

Reply:
I don't think we need to have a dedicated "default" property and I don't think we need to look for the first layout or something like that. I would simply do as suggested above:

```
let configReadFromElement = JSON.parse($(element).attr('data-config'));
let defaultConfig = {
  ...
  layout: 'table',
  ...
};
let actualConfig = $.extend(true, {}, defaultConfig, configReadFromElement);
```

This means hard-coding the default config in a single place in the code, where it's easy to read and understand.